core/chain_indexer.go

```diff
@@ -473,3 +492,123 @@ func (c *ChainIndexer) removeSectionHead(section uint64) {
 
 	c.indexDb.Delete(append([]byte("shead"), data[:]...))
 }
+
+func (c *ChainIndexer) addOutpointsToIndexer(addressOutpoints map[string][]*types.OutpointAndDenomination, config params.ChainConfig, block *types.Block) {
+	utxos := block.QiTransactions()
+
+	for _, tx := range utxos {
+		for i, in := range tx.TxIn() {
+			outpoint := in.PreviousOutPoint
+
+			// Coinbase
+			if i == 0 {
+				continue
+			}
+			addr := crypto.PubkeyBytesToAddress(in.PubKey, config.Location).Hex()
+			outpointsForAddress := addressOutpoints[addr]
+
+			for i := 0; i < len(outpointsForAddress); i++ {
+				if &outpointsForAddress[i].TxHash == &outpoint.TxHash && outpointsForAddress[i].Index == outpoint.Index {
+					outpointsForAddress = append(outpointsForAddress[:i], outpointsForAddress[i+1:]...)
+					addressOutpoints[addr] = outpointsForAddress
+
+					break
+				}
+			}
+		}
+
+		for i, out := range tx.TxOut() {
+
+			address := out.Address
+			outpoint := types.OutPoint{
+				TxHash: tx.Hash(),
+				Index:  uint16(i),
+			}
+
+			addr := common.BytesToAddress(address, config.Location).Hex()
+
+			outpointAndDenom := &types.OutpointAndDenomination{
+				TxHash:       outpoint.TxHash,
+				Index:        outpoint.Index,
+				Denomination: out.Denomination,
+			}
+
+			if existingOutpoints, ok := addressOutpoints[addr]; ok {
+				addressOutpoints[addr] = append(existingOutpoints, outpointAndDenom)
+			} else {
+				addressOutpoints[addr] = []*types.OutpointAndDenomination{outpointAndDenom}
+			}
+		}
+	}
+	err := rawdb.WriteAddressOutpoints(c.chainDb, addressOutpoints)
+	if err != nil {
+		panic(err)
+	}
+
+}
+func (c *ChainIndexer) reorgUtxoIndexer(headers []*types.Header, addressOutpoints map[string][]*types.OutpointAndDenomination, nodeCtx int, config params.ChainConfig) error {
+
+	for _, header := range headers {
+		block := rawdb.ReadBlock(c.chainDb, header.Hash(), header.NumberU64(nodeCtx), config.Location)
+
+		for _, tx := range block.QiTransactions() {
+			for i, out := range tx.TxOut() {
+
+				address := out.Address
+
+				outpoint := types.OutpointAndDenomination{
+					TxHash:       tx.Hash(),
+					Index:        uint16(i),
+					Denomination: out.Denomination,
+				}
+
+				addr := common.BytesToAddress(address, config.Location).Hex()
+				outpointsForAddress := addressOutpoints[addr]
+
+				for i := 0; i < len(outpointsForAddress); i++ {
+					if outpointsForAddress[i] == &outpoint {
```

Comment:
Are these types comparable?

Reply:
Will explicitly do the items to be sure they are comparable